File: zerocool-kor/valid-parentheses.go
package main

import (
	"log"
)

type ParenthesesStack []int8

func (s *ParenthesesStack) IsEmpty() bool {
	return len(*s) == 0
}

func (s *ParenthesesStack) Push(ele int8) {
	*s = append(*s, ele)
}

func (s *ParenthesesStack) Pop() (int8, bool) {
	if s.IsEmpty() {
		return 0, false
	} else {
		index := len(*s) - 1
		element := (*s)[index]
		*s = (*s)[:index]
		return element, true
	}
}

var openParentheses = map[string]bool{
	"(": true,
	"{": true,
	"[": true,
}
var pmap = map[string]int8{
	"(": 1,
	"{": 2,
	"[": 3,
	")": -1,
	"}": -2,
	"]": -3,
}

func isValid(s string) bool {
	log.Println("isValid(", s, ") started")
	var openList ParenthesesStack
	for _, ch := range s {
		letter := string(ch)
		//log.Println("i:",i,", letter:", letter)
		n, found := pmap[letter]
		if !found {
			return false
		}
		_, open := openParentheses[letter]
		if open {
			openList.Push(n)
		} else {
			poped, _ := openList.Pop()
			if poped+n != 0 {
				return false
			}
		}
	}
	return true
}

func main() {
	log.Println("result -> ", isValid("abcd"))
	log.Println("result -> ", isValid("()"))
	log.Println("result -> ", isValid("(){}[]({[]})"))

}

File: zerocool-kor/leetcode/valid-parentheses.ts
const isValid = (s:string): boolean => {
    const openParentheses = ['(','{','['];
    const pmap = {
        '(': 1,
        '{' : 2,
        '[' : 3,
        ']' : -3,
        '}' : -2,
        ')' : -1
    }
    console.log(`isValid(${s}) started`);
    const openList = [];
    for(const c of s.split('')){
        const n = pmap[c];
        if(n===undefined){
            return false;
        }
        if(openParentheses.indexOf(c)>=0){
            openList.push(n);       
        }else if(openList.pop() + n !== 0){
            return false;
        }       
    }
    return true;
}

console.log('result -> ' + isValid('abcd'));
console.log('result -> ' + isValid('()'));
console.log('result -> ' + isValid('(){}[]({[]})'));

